src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -371,16 +373,25 @@ InputOrderInfoPtr buildInputOrderInfo(
     size_t next_description_column = 0;
     size_t next_sort_key = 0;
 
-    while (next_description_column < description.size() && next_sort_key < sorting_key_columns.size())
+    while (next_description_column < description.size() && next_sort_key < sorting_key.column_names.size())
     {
-        const auto & sorting_key_column = sorting_key_columns[next_sort_key];
+        const auto & sorting_key_column = sorting_key.column_names[next_sort_key];
         const auto & sort_column_description = description[next_description_column];
 
         /// If required order depend on collation, it cannot be matched with primary key order.
         /// Because primary keys cannot have collations.
         if (sort_column_description.collator)
             break;
 
+        /// Since sorting key columns are always sorted with NULLS LAST, reading in order
+        /// supported only for ASC NULLS LAST ("in order"), and DESC NULLS FIRST ("reverse")
+        const auto column_is_nullable = sorting_key.data_types[next_sort_key]->isNullable();
+        if (column_is_nullable
+            && !(
+                (sort_column_description.direction == 1 && sort_column_description.nulls_direction == 1)
+                || (sort_column_description.direction == -1 && sort_column_description.nulls_direction == 1)))
```

Comment:
Can this condition be more readable?
```
column_is_nullable && 
sort_column_description.nulls_direction != 1 && 
sort_column_description.direction != sort_column_description.nulls_direction
```

Reply:
Done (null ordering check is enough) -- initially decided to make this check as straightforward as possible.